### PR TITLE
New Rule: Always use `true` as the final condition of a `cond` statement

### PR DESCRIPTION
**Pull request by 5thWall**
From the community style guide.

### CODE REVIEW

File: lib/dogma/rules/final_condition.ex [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  end
+
+  defp check_node(node, errors, _) do
+    {node, errors}
+  end
+
+  defp find_last_cond([[do: children] | _]) do
+    children
+    |> Enum.reverse
+    |> hd
```
Comment (jacknoble) [inline lib/dogma/rules/final_condition.ex]: List.last/1 ?

**Comment by lpil** [comment]
I like that we have this rule, but I'm not sure I like using true as the catch all. It doesn't read overly well, and it's not in keeping with the Elixir standard of using `_` as the catch all.

**Comment by meadsteve** [comment]
Interestingly this community style guide recommends using true as the catch all http://elixir.community/styleguide however I agree with @lpil and think it doesn't read very well.

**Comment by lpil** [comment]
`_` works as a catch all in a `cond`, correct? If so, I think we should recommend that as it is the same as `case` and pattern matching function definitions.

I am also rather partial to `:otherwise`, though that's probably the Haskell experience talking.

File: lib/dogma/rules/final_condition.ex (hunk 1 above)
Comment (lpil) [inline lib/dogma/rules/final_condition.ex]: Thanks :)

**Comment by meadsteve** [comment]
_ definitely works. And it's what I'd prefer but the defaults for this project should really be the accepted community standard (if one exists). In this case I'm not actually sure whether it's ```true``` or ```_```

**Comment by lpil** [comment]
Looking at elixir-lang/ex_doc it seems to be `true`.

What should this rule do when we explicitly do not want a catch all?

**Comment by 5thWall** [comment]
`_` is the catch all for pattern matching, but there's nothing to match against in a `cond` statement. It doesn't really make sense to me why it works at all, except just to match `case` statements.

For when we don't want a catch all, I'm actually not sure. The community styleguide just says that the final condition should always be true. `cond` statements will fail at runtime if at least one statement doesn't evaluate to truthy. So this could guard against that, but maybe someone want's to fail fast.

**Comment by lpil** [comment]
> cond statements will fail at runtime if at least one statement doesn't evaluate to truthy. So this could guard against that, but maybe someone want's to fail fast.

That's kind of the aim of the game with Elixir and Erlang in my opinion!

**Comment by 5thWall** [comment]
True. I could set it up so that it checks to see if the final condition is a literal value, and only returns an error then. That way it would skip over statements where the final condition is a method call or other expression.

**Comment by meadsteve** [comment]
At the risk of overcomplicating our config we could make it an option whether a catch-all is required or not.

**Comment by lpil** [comment]
I disagree, it should be considered on a case by case basis.

Checking checking if it is a literal or not sounds good. We do have a module that checks whether an AST node is a literal or not.

**Comment by 5thWall** [comment]
Okay, I'll add checking to see if the last value is a literal and compare there. Where are we on what the default should be?

**Comment by lpil** [comment]
I vote `true`, as that seems to be what Jose uses.

File: docs/rules.md [hunk 2]
```diff
@@ -75,6 +76,67 @@ A rule that disallows calls to IEx.pry, as while useful, we probably don't
...
+      1 + 2 == 5 ->
+        "Nope"
+      1 + 3 == 5 ->
+        "Nada"
+      _ ->
+        "OK"
+    end
+
+This rule will only catch those `cond` statements where the last condition
+is a literall or a `_`. Complex expressions and function calls will not
```
Comment (lpil) [inline docs/rules.md]: Typo, 'literal'

**Comment by lpil** [comment]
Nice work. Happy to merge after the typo fix.

**Pull request merged**